### PR TITLE
[mlir][evm] Initial impl of bytecode generation

### PR DESCRIPTION
**Pull request by abinavpp**

### CODE REVIEW

File: libsolidity/codegen/mlir/Passes.cpp [hunk 1]
```diff
@@ -212,7 +212,43 @@ static mlir::ModuleOp extractRuntimeModule(mlir::ModuleOp creationMod) {
...
+  outStreamData.clear();
+  llvmPassMgr.run(runtimeMod);
+  objs[1] = LLVMCreateMemoryBufferWithMemoryRange(
+      outStream.str().data(), outStream.str().size(), "Input",
+      /*RequiresNullTerminator=*/0);
+  objIds[1] = "Test_deployed"; // FIXME
+
+  LLVMMemoryBufferRef bytecodes[2];
+  char *errMsg = nullptr;
+  if (LLVMLinkEVM(objs, objIds, /*numInBuffers=*/2, bytecodes, &errMsg))
```
Comment (abinavpp) [inline libsolidity/codegen/mlir/Passes.cpp:237]: @PavelKopyl This is failing at LinkerDriver::linkerMain, but I can't see the error message. Do you see anything wrong in the usage here?

**Review by hedgar2017 — APPROVED** [review]

**Review by PavelKopyl — APPROVED** [review]
LGTM.

File: libsolidity/codegen/mlir/Passes.cpp (hunk 1 above)
Comment (PavelKopyl) [reply libsolidity/codegen/mlir/Passes.cpp:237]: So, the `errMsg` is empty?

Comment (abinavpp) [reply libsolidity/codegen/mlir/Passes.cpp:237]: I tracked down the error. It's happening here: https://github.com/matter-labs/era-compiler-llvm/blob/main/llvm/include/llvm/Object/ELF.h#L876
Do you know what's causing this?

Also, do you know how I can get this error message in `errMsg`. The llvm_unreachable under this is not executed in this case.

Comment (PavelKopyl) [reply libsolidity/codegen/mlir/Passes.cpp:237]: Hm.. it's strange that it's not executed. It looks like we passes to the linker a broken elf. 
By the way, let's rewrite the code:
```
  llvm::legacy::PassManager llvmPassMgr;
  llvm::SmallString<0> outStreamData;
  llvm::raw_svector_ostream outStream(outStreamData);
  tgtMach.addPassesToEmitFile(llvmPassMgr, outStream,
                              /*DwoOut=*/nullptr,
                              llvm::CodeGenFileType::CGFT_ObjectFile);

  LLVMMemoryBufferRef objs[2];

  // Generate creation obj.
  llvmPassMgr.run(creationMod);
  objs[0] = LLVMCreateMemoryBufferWithMemoryRange(
      outStream.str().data(), outStream.str().size(), "Input",
      /*RequiresNullTerminator=*/0);
  objIds[0] = "Test"; // FIXME
```
to something like:
```
LLVMTargetMachineRef TMRef = wrap(tgtMach);
LLVMModuleRef MRef = wrap(creationMod);
char *ErrorMsg = 0; 
if (LLVMTargetMachineEmitToMemoryBuffer(TMRef, MRef, LLVMObjectFile, &ErrorMsg,
                                          &objs[0]))
    llvm_unreachable("Error");
```

And the same for the runtime module.

Comment (abinavpp) [reply libsolidity/codegen/mlir/Passes.cpp:237]: Thank you! I'm now able to see the error from lld, which is:

ld.lld: error: undefined symbol: __datasize_C_25_deployed
>>> referenced by LLVMDialectModule
>>>               Test:(.text+0x13)

ld.lld: error: undefined symbol: __dataoffset_C_25_deployed
>>> referenced by LLVMDialectModule
>>>               Test:(.text+0x18)

File: libsolidity/codegen/mlir/Passes.cpp [hunk 2]
```diff
@@ -209,10 +209,40 @@ static mlir::ModuleOp extractRuntimeModule(mlir::ModuleOp creationMod) {
   return {};
 }
 
+DEFINE_SIMPLE_CONVERSION_FUNCTIONS(llvm::TargetMachine, LLVMTargetMachineRef)
```
Comment (abinavpp) [inline libsolidity/codegen/mlir/Passes.cpp:212]: @PavelKopyl Did I miss this in the llvm headers?

File: libsolidity/codegen/mlir/Passes.cpp (hunk 1 above)
Comment (PavelKopyl) [reply libsolidity/codegen/mlir/Passes.cpp:237]: The names passed in `objIds` have to correspond to the names of objects in the Yul representation.
I suppose in the sol test that you have compiled, the object names are: "C_25" - top level object; "C_25_deployed" - runtime object.

Comment (abinavpp) [reply libsolidity/codegen/mlir/Passes.cpp:237]: Sorry, this was due to the incorrect hardcoded inBufferIDs args to LLVMLinkEVM. I'm able to generate the bytecode when I fix that. Thanks for all the help! I'll close this PR as I'm able to generate the bytecode. I'll need to fix the tracking of yul object ids in the pipeline, which should be a separate work.

File: libsolidity/codegen/mlir/Passes.cpp (hunk 2 above)
Comment (PavelKopyl) [reply libsolidity/codegen/mlir/Passes.cpp:212]: As I understand it should be defined in `include/llvm/Support/CBindingWrapping.h`

Comment (abinavpp) [reply libsolidity/codegen/mlir/Passes.cpp:212]: Sorry I was not clear. I meant if I missed the header that does `DEFINE_SIMPLE_CONVERSION_FUNCTIONS(llvm::TargetMachine, LLVMTargetMachineRef)`?
I guess not, as I can't find it in TargetMachine.h and I can see the same thing in, for instance, lib/ExecutionEngine/Orc/OrcV2CBindings.cpp:107

**Pull request closed without merging**